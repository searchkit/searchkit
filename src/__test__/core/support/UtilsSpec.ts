import {Utils} from "../../../"


describe("Utils", ()=> {


  it("guid()", ()=> {
    expect(Utils.guid().length).toEqual(36)
  })

  it("collapse()", ()=> {
    let times2 = n => n*2
    expect(
      Utils.collapse([times2, times2, times2], 1)
    ).toBe(8)
  })

  it("instanceOf", ()=> {
    let isRegex = Utils.instanceOf(RegExp)
    expect(isRegex(/s/)).toBe(true)
    expect(isRegex("s")).toBe(false)
  })

  it("interpolate", ()=> {
    expect(Utils.interpolate("hello {message}, Hello {message}{ending}{missing}", {
      message:"World",
      ending:"!"
    })).toEqual("hello World, Hello World!{missing}")
  })

  it("translate() - don't call interpolate", ()=> {
    let key = "No Results found!"
    spyOn(Utils, "interpolate")
    expect(Utils.translate(key)).toBe(key)
    expect(Utils.interpolate).not.toHaveBeenCalled()
  })

  it("translate() - with interpolations", ()=> {
    let key = "{count} Results found!"
    spyOn(Utils, "interpolate").and.callThrough()
    expect(Utils.translate(key, {count:5})).toBe(
      "5 Results found!"
    )
    expect(Utils.interpolate).toHaveBeenCalledWith(
      key, {count:5}
    )
  })

  it("autobind()", ()=> {
    class Foo {
      message="hello"

      @Utils.autobind
      getMessage(something){
        return this.message + " " + something
      }
    }

    let fn = new Foo().getMessage
    expect(fn("world")).toBe("hello world")

  })
})
