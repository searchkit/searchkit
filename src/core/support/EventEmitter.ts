import {without} from "lodash"
import {each} from "lodash"

export class EventEmitter {
  listeners = []

  addListener(fn){
    this.listeners.push(fn)
    return ()=>{
      this.listeners = without(this.listeners, fn)
    }
  }

  trigger(...args){
    each(this.listeners, (fn)=> {
      fn.apply(null, args)
    })
  }
}
