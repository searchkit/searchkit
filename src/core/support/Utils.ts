import {reduce} from "lodash"
import {map} from "lodash"
import {reject} from "lodash"
import {isUndefined} from "lodash"

export class Utils {
  static guidCounter = 0

  static guid(prefix=""){
    let id = ++Utils.guidCounter
    return prefix.toString() + id
  }

  static collapse(collection, seed){
    const reducer = (current, fn)=> fn(current)
    return reduce(collection, reducer, seed)
  }

  static instanceOf(klass){
    return (val)=> val instanceof klass
  }

  static interpolate(str, interpolations){
    return str.replace(
  		/{([^{}]*)}/g,
  		(a, b) => {
  			var r = interpolations[b];
  			return typeof r === 'string' || typeof r === 'number' ? r : a;
  		}
    )
  }

  static translate(key, interpolations?){
    if(interpolations){
      return Utils.interpolate(key, interpolations)
    } else {
      return key
    }
  }

  static computeOptionKeys(options, fields, defaultKey){
    return map(options, (option)=> {
      return Utils.generateKeyFromFields(option, fields, defaultKey)
    })
  }

  static generateKeyFromFields(ob, fields, defaultKey){
    if(ob.key){
      return ob
    }
    let fieldValues = reject(map(fields, (field: string)=> ob[field]), isUndefined)
    if(fieldValues.length > 0){
      ob.key = fieldValues.join("_")
    } else {
      ob.key = defaultKey
    }
    return ob
  }
}
