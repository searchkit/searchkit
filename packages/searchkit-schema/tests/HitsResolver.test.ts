import nock from 'nock'
import { gql } from 'apollo-server-micro'
import { MultiMatchQuery, SearchkitConfig } from '@searchkit/sdk'
import { setupTestServer, callQuery } from './support/helper'
import HitsMock from './__mock-data__/HitResolver/Hits.json'
import HighlightHitsMock from './__mock-data__/HitResolver/HighlightHits.json'

describe('Hits Resolver', () => {
  describe('should return as expected', () => {
    const runQuery = async (
      query = '',
      page = { size: 10, from: 0 },
      sorting?: string,
      includeHighlight?: boolean
    ) => {
      const gql = `
        {
          results(query: "${query}") {
            hits(page: {size: ${page.size}, from: ${page.from}} ${
        sorting ? `, sortBy: "${sorting}"` : ''
      }) {
              sortedBy
              items {
                ... on ResultHit {
                  id
                  fields {
                    writers
                    actors
                  }
                  ${includeHighlight ? 'highlight' : ''}
                }
              }
            }
          }
        }
      `
      const response = await callQuery({ gql })
      return response
    }

    it('should return correct Results', async () => {
      const config: SearchkitConfig = {
        host: 'http://localhost:9200',
        index: 'movies',
        hits: {
          fields: ['actors', 'writers']
        },
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {},
              "from": 0,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery()
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('should return correct Results with https', async () => {
      const config: SearchkitConfig = {
        host: 'https://localhost:9200',
        index: 'movies',
        hits: {
          fields: ['actors', 'writers']
        },
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const scope = nock('https://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "aggs": Object {},
              "from": 0,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery()
      expect(response.status).toEqual(200)
    })

    it('should return correct Results on page 2', async () => {
      const config: SearchkitConfig = {
        host: 'http://localhost:9200',
        index: 'movies',
        hits: {
          fields: ['actors', 'writers']
        },
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {},
              "from": 10,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const response = await runQuery('', { size: 10, from: 10 })
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('should provide sorting options', async () => {
      const config: SearchkitConfig = {
        host: 'http://localhost:9200',
        index: 'movies',
        hits: {
          fields: ['actors', 'writers']
        },
        sortOptions: [
          { id: 'relevance', label: 'Relevance', field: '_score' },
          { id: 'released', label: 'Recent Releases', field: { released: 'desc' } }
        ],
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {},
              "from": 10,
              "size": 10,
              "sort": "_score",
            }
          `)
          return [200, HitsMock]
        })

      let response = await runQuery('', { size: 10, from: 10 }, 'relevance')
      expect(response.body.data).toMatchSnapshot()

      scope.post('/movies/_search').reply((uri, body) => {
        expect(body).toMatchInlineSnapshot(`
          Object {
            "_source": Object {
              "includes": Array [
                "actors",
                "writers",
              ],
            },
            "aggs": Object {},
            "from": 10,
            "size": 10,
            "sort": Object {
              "released": "desc",
            },
          }
        `)
        return [200, HitsMock]
      })

      response = await runQuery('', { size: 10, from: 10 }, 'released')
      expect(response.body.data).toMatchSnapshot()

      expect(response.status).toEqual(200)
    })

    it('should return correct highlight Results', async () => {
      const config: SearchkitConfig = {
        host: 'http://localhost:9200',
        index: 'movies',
        hits: {
          fields: ['actors', 'writers'],
          highlightedFields: [
            'actors',
            { field: 'writers', config: { pre_tags: ['<b>'], post_tags: ['</b>'] } }
          ]
        },
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      setupTestServer(
        {
          config,
          addToQueryType: true,
          typeName: 'ResultSet',
          hitTypeName: 'ResultHit'
        },
        gql`
          type Query {
            root: String
          }

          type Mutation {
            root: String
          }

          type HitFields {
            title: String
            writers: [String]
            actors: [String]
            plot: String
            poster: String
          }

          type ResultHit implements SKHit {
            id: ID!
            fields: HitFields
            highlight: String
          }
        `,
        {
          ResultHit: {
            highlight: (hit: any) => JSON.stringify(hit.highlight)
          }
        }
      )

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {},
              "from": 0,
              "highlight": Object {
                "fields": Object {
                  "actors": Object {},
                  "writers": Object {
                    "post_tags": Array [
                      "</b>",
                    ],
                    "pre_tags": Array [
                      "<b>",
                    ],
                  },
                },
              },
              "query": Object {
                "bool": Object {
                  "should": Array [
                    Object {
                      "multi_match": Object {
                        "fields": Array [
                          "actors",
                          "writers",
                          "title^4",
                          "plot",
                        ],
                        "operator": "and",
                        "query": "al",
                        "type": "best_fields",
                      },
                    },
                    Object {
                      "multi_match": Object {
                        "fields": Array [
                          "actors",
                          "writers",
                          "title^4",
                          "plot",
                        ],
                        "query": "al",
                        "type": "cross_fields",
                      },
                    },
                    Object {
                      "multi_match": Object {
                        "fields": Array [
                          "actors",
                          "writers",
                          "title^4",
                          "plot",
                        ],
                        "query": "al",
                        "type": "phrase",
                      },
                    },
                    Object {
                      "multi_match": Object {
                        "fields": Array [
                          "actors",
                          "writers",
                          "title^4",
                          "plot",
                        ],
                        "query": "al",
                        "type": "phrase_prefix",
                      },
                    },
                  ],
                },
              },
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HighlightHitsMock]
        })

      const response = await runQuery('al', { size: 10, from: 0 }, undefined, true)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
