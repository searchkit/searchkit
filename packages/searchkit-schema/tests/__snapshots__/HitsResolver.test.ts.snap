// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits Resolver should return as expected should provide sorting options 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Christian Bale",
              "Heath Ledger",
              "Aaron Eckhart",
              "Michael Caine",
            ],
            "writers": Array [
              "Jonathan Nolan",
              "Christopher Nolan",
              "David S. Goyer",
              "Bob Kane",
            ],
          },
          "id": "tt0468569",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Tim Roth",
              "Amanda Plummer",
              "Laura Lovelace",
              "John Travolta",
            ],
            "writers": Array [
              "Quentin Tarantino",
              "Roger Avary",
            ],
          },
          "id": "tt0110912",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Noel Appleby",
              "Alexandra Astin",
              "Sean Astin",
              "David Aston",
            ],
            "writers": Array [
              "J.R.R. Tolkien",
              "Fran Walsh",
              "Philippa Boyens",
              "Peter Jackson",
            ],
          },
          "id": "tt0167260",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Morgan Freeman",
              "Andrew Kevin Walker",
              "Daniel Zacapa",
              "Brad Pitt",
            ],
            "writers": Array [
              "Andrew Kevin Walker",
            ],
          },
          "id": "tt0114369",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Michael Berryman",
              "Peter Brocco",
              "Dean R. Brooks",
              "Alonzo Brown",
            ],
            "writers": Array [
              "Lawrence Hauben",
              "Bo Goldman",
              "Ken Kesey",
              "Dale Wasserman",
            ],
          },
          "id": "tt0073486",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Russell Crowe",
              "Joaquin Phoenix",
              "Connie Nielsen",
              "Oliver Reed",
            ],
            "writers": Array [
              "David Franzoni",
              "John Logan",
              "William Nicholson",
            ],
          },
          "id": "tt0172495",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Andrew Lincoln",
              "Steven Yeun",
              "Chandler Riggs",
              "Norman Reedus",
            ],
            "writers": Array [
              "Frank Darabont",
            ],
          },
          "id": "tt1520211",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Carrey",
              "Kate Winslet",
              "Gerry Robert Byrne",
              "Elijah Wood",
            ],
            "writers": Array [
              "Charlie Kaufman",
              "Michel Gondry",
              "Pierre Bismuth",
            ],
          },
          "id": "tt0338013",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Marlon Brando",
              "Martin Sheen",
              "Robert Duvall",
              "Frederic Forrest",
            ],
            "writers": Array [
              "John Milius",
              "Francis Ford Coppola",
              "Michael Herr",
            ],
          },
          "id": "tt0078788",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Ellen Burstyn",
              "Jared Leto",
              "Jennifer Connelly",
              "Marlon Wayans",
            ],
            "writers": Array [
              "Hubert Selby Jr.",
              "Darren Aronofsky",
            ],
          },
          "id": "tt0180093",
        },
      ],
      "sortedBy": "relevance",
    },
  },
}
`;

exports[`Hits Resolver should return as expected should provide sorting options 4`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Christian Bale",
              "Heath Ledger",
              "Aaron Eckhart",
              "Michael Caine",
            ],
            "writers": Array [
              "Jonathan Nolan",
              "Christopher Nolan",
              "David S. Goyer",
              "Bob Kane",
            ],
          },
          "id": "tt0468569",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Tim Roth",
              "Amanda Plummer",
              "Laura Lovelace",
              "John Travolta",
            ],
            "writers": Array [
              "Quentin Tarantino",
              "Roger Avary",
            ],
          },
          "id": "tt0110912",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Noel Appleby",
              "Alexandra Astin",
              "Sean Astin",
              "David Aston",
            ],
            "writers": Array [
              "J.R.R. Tolkien",
              "Fran Walsh",
              "Philippa Boyens",
              "Peter Jackson",
            ],
          },
          "id": "tt0167260",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Morgan Freeman",
              "Andrew Kevin Walker",
              "Daniel Zacapa",
              "Brad Pitt",
            ],
            "writers": Array [
              "Andrew Kevin Walker",
            ],
          },
          "id": "tt0114369",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Michael Berryman",
              "Peter Brocco",
              "Dean R. Brooks",
              "Alonzo Brown",
            ],
            "writers": Array [
              "Lawrence Hauben",
              "Bo Goldman",
              "Ken Kesey",
              "Dale Wasserman",
            ],
          },
          "id": "tt0073486",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Russell Crowe",
              "Joaquin Phoenix",
              "Connie Nielsen",
              "Oliver Reed",
            ],
            "writers": Array [
              "David Franzoni",
              "John Logan",
              "William Nicholson",
            ],
          },
          "id": "tt0172495",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Andrew Lincoln",
              "Steven Yeun",
              "Chandler Riggs",
              "Norman Reedus",
            ],
            "writers": Array [
              "Frank Darabont",
            ],
          },
          "id": "tt1520211",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Carrey",
              "Kate Winslet",
              "Gerry Robert Byrne",
              "Elijah Wood",
            ],
            "writers": Array [
              "Charlie Kaufman",
              "Michel Gondry",
              "Pierre Bismuth",
            ],
          },
          "id": "tt0338013",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Marlon Brando",
              "Martin Sheen",
              "Robert Duvall",
              "Frederic Forrest",
            ],
            "writers": Array [
              "John Milius",
              "Francis Ford Coppola",
              "Michael Herr",
            ],
          },
          "id": "tt0078788",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Ellen Burstyn",
              "Jared Leto",
              "Jennifer Connelly",
              "Marlon Wayans",
            ],
            "writers": Array [
              "Hubert Selby Jr.",
              "Darren Aronofsky",
            ],
          },
          "id": "tt0180093",
        },
      ],
      "sortedBy": "released",
    },
  },
}
`;

exports[`Hits Resolver should return as expected should return correct Results 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Christian Bale",
              "Heath Ledger",
              "Aaron Eckhart",
              "Michael Caine",
            ],
            "writers": Array [
              "Jonathan Nolan",
              "Christopher Nolan",
              "David S. Goyer",
              "Bob Kane",
            ],
          },
          "id": "tt0468569",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Tim Roth",
              "Amanda Plummer",
              "Laura Lovelace",
              "John Travolta",
            ],
            "writers": Array [
              "Quentin Tarantino",
              "Roger Avary",
            ],
          },
          "id": "tt0110912",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Noel Appleby",
              "Alexandra Astin",
              "Sean Astin",
              "David Aston",
            ],
            "writers": Array [
              "J.R.R. Tolkien",
              "Fran Walsh",
              "Philippa Boyens",
              "Peter Jackson",
            ],
          },
          "id": "tt0167260",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Morgan Freeman",
              "Andrew Kevin Walker",
              "Daniel Zacapa",
              "Brad Pitt",
            ],
            "writers": Array [
              "Andrew Kevin Walker",
            ],
          },
          "id": "tt0114369",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Michael Berryman",
              "Peter Brocco",
              "Dean R. Brooks",
              "Alonzo Brown",
            ],
            "writers": Array [
              "Lawrence Hauben",
              "Bo Goldman",
              "Ken Kesey",
              "Dale Wasserman",
            ],
          },
          "id": "tt0073486",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Russell Crowe",
              "Joaquin Phoenix",
              "Connie Nielsen",
              "Oliver Reed",
            ],
            "writers": Array [
              "David Franzoni",
              "John Logan",
              "William Nicholson",
            ],
          },
          "id": "tt0172495",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Andrew Lincoln",
              "Steven Yeun",
              "Chandler Riggs",
              "Norman Reedus",
            ],
            "writers": Array [
              "Frank Darabont",
            ],
          },
          "id": "tt1520211",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Carrey",
              "Kate Winslet",
              "Gerry Robert Byrne",
              "Elijah Wood",
            ],
            "writers": Array [
              "Charlie Kaufman",
              "Michel Gondry",
              "Pierre Bismuth",
            ],
          },
          "id": "tt0338013",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Marlon Brando",
              "Martin Sheen",
              "Robert Duvall",
              "Frederic Forrest",
            ],
            "writers": Array [
              "John Milius",
              "Francis Ford Coppola",
              "Michael Herr",
            ],
          },
          "id": "tt0078788",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Ellen Burstyn",
              "Jared Leto",
              "Jennifer Connelly",
              "Marlon Wayans",
            ],
            "writers": Array [
              "Hubert Selby Jr.",
              "Darren Aronofsky",
            ],
          },
          "id": "tt0180093",
        },
      ],
      "sortedBy": null,
    },
  },
}
`;

exports[`Hits Resolver should return as expected should return correct Results on page 2 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Christian Bale",
              "Heath Ledger",
              "Aaron Eckhart",
              "Michael Caine",
            ],
            "writers": Array [
              "Jonathan Nolan",
              "Christopher Nolan",
              "David S. Goyer",
              "Bob Kane",
            ],
          },
          "id": "tt0468569",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Tim Roth",
              "Amanda Plummer",
              "Laura Lovelace",
              "John Travolta",
            ],
            "writers": Array [
              "Quentin Tarantino",
              "Roger Avary",
            ],
          },
          "id": "tt0110912",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Noel Appleby",
              "Alexandra Astin",
              "Sean Astin",
              "David Aston",
            ],
            "writers": Array [
              "J.R.R. Tolkien",
              "Fran Walsh",
              "Philippa Boyens",
              "Peter Jackson",
            ],
          },
          "id": "tt0167260",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Morgan Freeman",
              "Andrew Kevin Walker",
              "Daniel Zacapa",
              "Brad Pitt",
            ],
            "writers": Array [
              "Andrew Kevin Walker",
            ],
          },
          "id": "tt0114369",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Michael Berryman",
              "Peter Brocco",
              "Dean R. Brooks",
              "Alonzo Brown",
            ],
            "writers": Array [
              "Lawrence Hauben",
              "Bo Goldman",
              "Ken Kesey",
              "Dale Wasserman",
            ],
          },
          "id": "tt0073486",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Russell Crowe",
              "Joaquin Phoenix",
              "Connie Nielsen",
              "Oliver Reed",
            ],
            "writers": Array [
              "David Franzoni",
              "John Logan",
              "William Nicholson",
            ],
          },
          "id": "tt0172495",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Andrew Lincoln",
              "Steven Yeun",
              "Chandler Riggs",
              "Norman Reedus",
            ],
            "writers": Array [
              "Frank Darabont",
            ],
          },
          "id": "tt1520211",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Carrey",
              "Kate Winslet",
              "Gerry Robert Byrne",
              "Elijah Wood",
            ],
            "writers": Array [
              "Charlie Kaufman",
              "Michel Gondry",
              "Pierre Bismuth",
            ],
          },
          "id": "tt0338013",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Marlon Brando",
              "Martin Sheen",
              "Robert Duvall",
              "Frederic Forrest",
            ],
            "writers": Array [
              "John Milius",
              "Francis Ford Coppola",
              "Michael Herr",
            ],
          },
          "id": "tt0078788",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Ellen Burstyn",
              "Jared Leto",
              "Jennifer Connelly",
              "Marlon Wayans",
            ],
            "writers": Array [
              "Hubert Selby Jr.",
              "Darren Aronofsky",
            ],
          },
          "id": "tt0180093",
        },
      ],
      "sortedBy": null,
    },
  },
}
`;

exports[`Hits Resolver should return as expected should return correct highlight Results 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "fields": Object {
            "actors": Array [
              "Tom Neal",
              "Ann Savage",
              "Claudia Drake",
              "Edmund MacDonald",
            ],
            "writers": Array [
              "Martin Goldsmith",
              "Martin Goldsmith",
            ],
          },
          "highlight": "{}",
          "id": "tt0037638",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Al Gore",
              "Billy West",
            ],
            "writers": Array [],
          },
          "highlight": "{\\"actors\\":[\\"<em>Al</em> Gore\\"]}",
          "id": "tt0497116",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Samir Khader",
              "Josh Rushing",
              "George W. Bush",
              "Hassan Ibrahim",
            ],
            "writers": Array [
              "Julia Bacha",
              "Jehane Noujaim",
            ],
          },
          "highlight": "{}",
          "id": "tt0391024",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Kevin Costner",
              "Sean Connery",
              "Charles Martin Smith",
              "Andy Garcia",
            ],
            "writers": Array [
              "Oscar Fraley",
              "David Mamet",
            ],
          },
          "highlight": "{}",
          "id": "tt0094226",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jon Lovitz",
              "Nancy Cartwright",
              "Christine Cavanaugh",
              "Gerrit Graham",
            ],
            "writers": Array [
              "Al Jean",
              "Mike Reiss",
            ],
          },
          "highlight": "{\\"writers\\":[\\"<b>Al</b> Jean\\"]}",
          "id": "tt0108734",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Penelope Allen",
              "Gordon MacDonald",
              "Madison Arnold",
              "Vincent Angell",
            ],
            "writers": Array [
              "William Shakespeare",
              "Al Pacino",
            ],
          },
          "highlight": "{\\"writers\\":[\\"<b>Al</b> Pacino\\"]}",
          "id": "tt0116913",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "'Weird Al' Yankovic",
              "Victoria Jackson",
              "Kevin McCarthy",
              "Michael Richards",
            ],
            "writers": Array [
              "'Weird Al' Yankovic",
              "Jay Levey",
            ],
          },
          "highlight": "{\\"actors\\":[\\"'Weird <em>Al</em>' Yankovic\\"],\\"writers\\":[\\"'Weird <b>Al</b>' Yankovic\\"]}",
          "id": "tt0098546",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Jim Lovell",
              "Russell Schweickart",
              "Eugene Cernan",
              "Michael Collins",
            ],
            "writers": Array [],
          },
          "highlight": "{}",
          "id": "tt0097372",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "Buster Keaton",
              "Marion Mack",
              "Glen Cavender",
              "Jim Farley",
            ],
            "writers": Array [
              "Buster Keaton",
              "Clyde Bruckman",
              "Al Boasberg",
            ],
          },
          "highlight": "{\\"writers\\":[\\"<b>Al</b> Boasberg\\"]}",
          "id": "tt0017925",
        },
        Object {
          "fields": Object {
            "actors": Array [
              "James Caan",
              "Billy Dee Williams",
              "Jack Warden",
              "Bernie Casey",
            ],
            "writers": Array [
              "Gale Sayers",
              "Al Silverman",
              "William Blinn",
            ],
          },
          "highlight": "{\\"writers\\":[\\"<b>Al</b> Silverman\\"]}",
          "id": "tt0068315",
        },
      ],
      "sortedBy": null,
    },
  },
}
`;
